frame/transaction-payment/asset-conversion-tx-payment/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+pub(crate) type AssetBalanceOf<T> =
+	<<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+/// Type alias for Asset IDs.
+pub(crate) type AssetIdOf<T> =
+	<<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+
+/// Type alias for the interaction of balances with `OnChargeAssetTransaction`.
+pub(crate) type ChargeAssetBalanceOf<T> =
+	<<T as Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::Balance;
+/// Type alias for Asset IDs in their interaction with `OnChargeAssetTransaction`.
+pub(crate) type ChargeAssetIdOf<T> =
+	<<T as Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::AssetId;
+/// Liquidity info type alias for interaction with `OnChargeAssetTransaction`.
+pub(crate) type ChargeAssetLiquidityOf<T> =
+	<<T as Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::LiquidityInfo;
+
+/// Used to pass the initial payment info from pre- to post-dispatch.
+#[derive(Encode, Decode, DefaultNoBound, TypeInfo)]
+pub enum InitialPayment<T: Config> {
+	/// No initial fee was paid.
+	#[default]
+	Nothing,
+	/// The initial fee was paid in the native currency.
+	Native(LiquidityInfoOf<T>),
+	/// The initial fee was paid in an asset.
+	Asset((LiquidityInfoOf<T>, BalanceOf<T>, AssetBalanceOf<T>)),
+}
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_transaction_payment::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The fungibles instance used to pay for transactions in assets.
+		type Fungibles: Balanced<Self::AccountId>;
+		/// The actual transaction charging logic that charges the fees.
+		type OnChargeAssetTransaction: OnChargeAssetTransaction<Self>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A transaction fee `actual_fee`, of which `tip` was added to the minimum inclusion fee,
+		/// has been paid by `who` in an asset `asset_id`.
+		AssetTxFeePaid {
+			who: T::AccountId,
+			actual_fee: AssetBalanceOf<T>,
+			tip: BalanceOf<T>,
+			asset_id: ChargeAssetIdOf<T>,
+		},
+	}
+}
+
+/// Require payment for transaction inclusion and optionally include a tip to gain additional
+/// priority in the queue. Allows paying via both `Currency` as well as `fungibles::Balanced`.
+///
+/// Wraps the transaction logic in [`pallet_transaction_payment`] and extends it with assets.
+/// An asset ID of `None` falls back to the underlying transaction payment logic via the native
+/// currency.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct ChargeAssetTxPayment<T: Config> {
+	#[codec(compact)]
+	tip: BalanceOf<T>,
+	asset_id: Option<ChargeAssetIdOf<T>>,
+}
```

Comment:
It's strange to me that the `tip` has type `BalanceOf` here (although I notice it's the same in the [other extension](https://github.com/paritytech/substrate/blob/js/pay-by-swap/frame/transaction-payment/asset-tx-payment/src/lib.rs#L152)). I would expect the tip also to be in an asset. I suppose this means you can only tip in the native asset directly?

I thought it was strange that the tip in the `Event` is also `BalanceOf`, while again the other extension [converts it](https://github.com/paritytech/substrate/blob/js/pay-by-swap/frame/transaction-payment/asset-tx-payment/src/lib.rs#L137). But this could be wrong if the tip isn't actually paid in `AssetBalance`, no?
```rust
AssetTxFeePaid {
	who: T::AccountId,
	actual_fee: AssetBalanceOf<T>,
	tip: BalanceOf<T>,
	asset_id: ChargeAssetIdOf<T>,
},
```

Reply:
Discussed with Jegor in DM, and just summarizing here. The tip is primarily used for prioritization, and of course from a collator's perspective, they just care about the native asset (`BalanceOf`). It gets added to the fee in pre-dispatch and converted. In post-dispatch, any refund is converted back to the asset, and the `actual_fee` is the total fee paid (`initial_inclusion_fee + tip - refund`). Since conversion back in post might not be at the same ratio as in pre (like it is with ratio of minimum balances), we'd be left guessing at what amount of the `actual_fee` is the tip. The main point is: The fee handler/collator got the amount of native asset needed.

I think in the other extension it was actually a mistake to list the tip in terms of the asset, and that this makes more sense.